### PR TITLE
Support ``:emphasize-lines:`` in PDF output (closes #1238)

### LINKED ISSUE #1238: Support :emphasize-lines: in PDF output
See [Pyramid Issue 667](https://github.com/Pylons/pyramid/issues/667)

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1238
- Originally reported by: Steve Piercy
- Originally created at: 2013-08-19T01:23:11.178

### PR DESCRIPTION
**Pull request by jfbu**
Brief testing seems ok, (also for literalinclude and for cases when the original code line is wrapped). I checked with background colour and frame colour.

In some PDF viewers there may be some artefacts at some zooming levels: very thin lines where  the rectangle areas for each successive line touch; this is known PDF viewer issue, there is nothing (easy) to do on LaTeX side.

So, new `VerbatimHighlightColor` key for `'sphinxsetup'` in `latex_elements` config dict.

Added a minimal test and updated docs.

To the extent #1238 is considered bug, this is bug-fix, so I based it on stable... but I have no firm opinion whether for 1.6.6 or 1.7 only.

### CODE REVIEW

File: sphinx/writers/latex.py
Comment (jfbu) [inline sphinx/writers/latex.py]: @mgeier I recall you had once some problem with the way we add Verbatim to body. Would

    self.body.append('\n' + hllines + '\n' + hlcode + '\n')

be ok for your extension nbsphinx ?

**Review by tk0miya — APPROVED** [review]

**Comment by jfbu** [comment]
Fixed these issues:

- original commit was active only under (default) `verbatimwrapslines=true`, not `=false`,

- original commit wasn't aware of line numbering shift from either `lineno-start` or `lineno-match`.

Now, behaviour for latex is like for html: the line numbers specified by `emphasize-lines` count the _actually selected lines_ (which may or may not match the line number printed near the code line in PDF or html output).

It is however hard to set up a test for this because this is realized by TeX macros, so one need to actually produce the PDF, it is not testable on the LaTeX file only.

If no objection I will merge this soon.

Thanks @tk0miya for first review!

File: doc/markup/code.rst
Comment (jfbu) [inline doc/markup/code.rst]: will remove that blank line after merge

**Comment by jfbu** [comment]
Again a commit, for this reason: if wrapping of too long lines in code-blocks was disabled by user, then the highlight only applies to the actual code, not to the whole line, like this:

capture d ecran 2017-12-10 a 17 06 29

with this it will look like this, which is also the output when too long lines are wrapped:

capture d ecran 2017-12-10 a 17 11 17

Notice that some html themes (for example the one for Sphinx own docs)  highlight like the first way above, but I find the second way at bottom is better; and anyway wrapping long code-lines is default, and doing like the first way would be more compilcated TeXnically with such wrapped code-lines.

File: sphinx/writers/latex.py
Comment (mgeier) [reply sphinx/writers/latex.py]: @jfbu Thanks for remembering this issue!

Yes, this should be fine, because the `SphinxVerbatim` environment is still contained in a single list element which is the last one appended to `self.body`.
It doesn't matter if `hllines` is a separate list item or if `hllines` and `hlcode` are concatenated into one list item.
For reference, this was the previous issue you were mentioning: https://github.com/spatialaudio/nbsphinx/issues/78 and #3224.

After that issue, I've started using `pushbody()` and `popbody()`, which makes the whole thing much more robust.
Since then, it doesn't matter at all how the partial strings are appended to `self.body`.

Comment (jfbu) [reply sphinx/writers/latex.py]: @mgeier thanks for explanations, I will push a commit using single list element formed from `hllines` and `hlcode`.

**Comment by jfbu** [comment]
I will merge after testing completes

**Comment by jfbu** [comment]
For the record, testing was done after some point only with ``:linenos:`` present, and 347f15c was buggy in its absence. This got fixed at 8de0abd (still for 1.6.6). Thanks @dietmarw for reporting.

**Pull request merged**
